### PR TITLE
[IMP] mail: replace discuss control panel by new top bar

### PR DESCRIPTION
**Pull request by seb-odoo**
Part of task-2369636

### CODE REVIEW

File: addons/mail/static/src/models/discuss/discuss.js [hunk 1]
```diff
@@ -202,7 +220,7 @@ function factory(dependencies) {
          * @param {string} newName
          */
         async renameThread(thread, newName) {
-            await this.async(() => thread.rename(newName));
+            await this.async(() => thread.setCustomName(newName));
```
Comment (alexkuhn) [inline addons/mail/static/src/models/discuss/discuss.js:223]: Would make sense to rename flow so that it uses "set custom name" over "rename", especially when the rename is motivated by increase of feature scope from introducing "renaming of channel" as a different feature.

File: addons/mail/static/src/models/thread_view/thread_viewer.js [hunk 2]
```diff
@@ -83,6 +83,12 @@ function factory(dependencies) {
         hasThreadView: attr({
             default: false,
         }),
+        /**
+         * Determines whether this thread viewer has a top bar.
+         */
+        hasTopbar: attr({
+            default: false,
+        }),
```
Comment (alexkuhn) [inline addons/mail/static/src/models/thread_view/thread_viewer.js:91]: For clarity, I would tell it determines whether this thread viewer _has the feature to possibly_ have a top bar.

I know in prior fields you made the difference between _"States"_ and _"Determines"_. IMO these were slightly less confusing, but still poor because the change of verb hints at a change of behaviour, not necessarily a change of actual meaning... Basically comment is quite un-declarative.
Basically my suggestion here formulates in a way the meaning of field is more declarative, thus defines field more clearly.

File: addons/mail/static/src/components/thread_view/thread_view.xml [hunk 3]
```diff
@@ -4,6 +4,9 @@
     <t t-name="mail.ThreadView" owl="1">
         <div class="o_ThreadView" t-att-data-correspondent-id="threadView and threadView.thread and threadView.thread.correspondent and threadView.thread.correspondent.id" t-att-data-thread-local-id="threadView and threadView.thread and threadView.thread.localId">
             <t t-if="threadView">
+                <t t-if="threadView.topbar">
+                    <ThreadViewTopbar class="border-bottom" localId="threadView.topbar.localId"/>
```
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/components/thread_view/thread_view.xml:8]: Nevermind, conditionally displayed. It's fine then

File: addons/mail/static/src/components/discuss_mobile_mailbox_selection/discuss_mobile_mailbox_selection.scss [hunk 4]
```diff
@@ -12,15 +12,7 @@
     padding: map-get($spacers, 2);
     z-index: 1;
 
-    &.o-active {
+    &.active {
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/components/discuss_mobile_mailbox_selection/discuss_mobile_mailbox_selection.scss:15]: Both classes can be kept if preferable.

File: addons/mail/static/src/components/discuss/discuss.js [hunk 5]
```diff
@@ -267,14 +262,22 @@ export class Discuss extends Component {
...
+        }
+        if (!isMailboxSelected) {
+            this.discuss.update({ thread: unlink() });
+        }
+        if (!isChatSelected) {
+            this.discuss.update({ isAddingChat: false });
+        }
+        if (!isChannelSelected) {
+            this.discuss.update({ isAddingChannel: false });
+        }
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/components/discuss/discuss.js:280]: I'll remove them

File: addons/mail/static/src/models/discuss/discuss.js (hunk 1 above)
Comment (seb-odoo) [reply addons/mail/static/src/models/discuss/discuss.js:223]: The flow is just the event handler in that case, I'll rename that method to it instead of its result.

File: addons/mail/static/src/models/thread_view/thread_viewer.js (hunk 2 above)
Comment (seb-odoo) [reply addons/mail/static/src/models/thread_view/thread_viewer.js:91]: But this field is not having the possibility the feature (as all thread views possibly have it), this field literally determines if the top bar is there or not.

Comment (alexkuhn) [reply addons/mail/static/src/models/thread_view/thread_viewer.js:91]: Oh ok, I misunderstand how this field is used. It would have been best if the value was set using field definition API rather than code flow, but that's actual due to the unfortunate workaround to decouple thread viewer and UI record at technical level (at the cost of less declarative code).

_**Let's ignore this problem for the time being.**_ We are aware of the issue and I think this code is reasonable while this field value is set at thread viewer creation, until we have proper support of model sharing with no identity crisis and also correct direction of code customization (i.e. discuss, chatter and chat window have to customize the thread viewer, not the other way around).

File: addons/mail/static/src/component_hooks/use_ref_to_model/use_ref_to_model.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ * or destroy).
+ *
+ * @param {Object} param0
+ * @param {string} param0.fieldName Name of the field on the target record.
+ * @param {string} param0.modelName Name of the model of the target record.
+ * @param {string} param0.localIdPropName Name of the prop of this component
+ *  containing the localId of the target record.
+ * @param {string} param0.refName Name of the t-ref on this component.
+ */
+export function useRefToModel({ fieldName, modelName, localIdPropName, refName }) {
```
Comment (alexkuhn) [inline addons/mail/static/src/component_hooks/use_ref_to_model/use_ref_to_model.js:19]: I suggest `propNameAsRecordLocalId`. It's more natural and less obscure than current name.

File: addons/mail/static/src/component_hooks/use_update_to_model/use_update_to_model.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+ * This hook provides support for binding the onMounted/onPatched hooks to the
+ * method of a target record.
+ *
+ * @param {Object} param0
+ * @param {string} param0.localIdPropName Name of the prop of this component
+ *  containing the localId of the target record.
+ * @param {string} param0.methodName Name of the method on the target record.
+ * @param {string} param0.modelName Name of the model of the target record.
+ */
+export function useUpdateToModel({ localIdPropName, methodName, modelName }) {
```
Comment (alexkuhn) [inline addons/mail/static/src/component_hooks/use_update_to_model/use_update_to_model.js:17]: `useUpdateToModel` and `useRefToModel` are strange and not elegant, but it goes in the right direction for putting everything in model.

Note that `useUpdateToModel` is assimilable to `afterUpdate`/`onX` dummy fields/lifecycle methods. But considering it's used to overcome DOM API that are not declarative enough (set focus + selection range), it's very much appropriate while using DOM element explicitly.

File: addons/mail/static/src/components/discuss_mobile_mailbox_selection/discuss_mobile_mailbox_selection.xml [hunk 8]
```diff
@@ -4,11 +4,9 @@
     <t t-name="mail.DiscussMobileMailboxSelection" owl="1">
         <div class="o_DiscussMobileMailboxSelection">
             <t t-foreach="orderedMailboxes" t-as="mailbox" t-key="mailbox.localId">
-                <button class="o_DiscussMobileMailboxSelection_button btn"
+                <button class="o_DiscussMobileMailboxSelection_button btn btn-secondary"
                     t-att-class="{
-                        'btn-primary': discuss.thread === mailbox,
-                        'btn-secondary': discuss.thread !== mailbox,
-                        'o-active': discuss.thread === mailbox,
+                        'active o-active': discuss.thread === mailbox,
```
Comment (alexkuhn) [inline addons/mail/static/src/components/discuss_mobile_mailbox_selection/discuss_mobile_mailbox_selection.xml:9]: cool for keeping `o-active`. Note that when all UI things will eventually be integrated to Models, the `o-active` would simply be a computed field... We could even just remove this as it's pretty much just an alias for `discuss.thread === mailbox`.

Reason to define it is because component logic and styling require introducing this classname, but it would no longer be necessary when all is model ;)

**Comment by alexkuhn** [comment]
Looks good in use. Just small remark about caret position on click thread name.
Considering the feedback cursor on hover is caret-shaped, would expect position of caret is very close to where we click.
Instead it always goes to the end.

click-always-put-caret-at-end

Maybe a bit complex to target proper position, so can ignore. Or maybe have a clickable-shaped cursor instead of caret-shaped?

File: addons/mail/static/src/component_hooks/use_ref_to_model/use_ref_to_model.js (hunk 6 above)
Comment (seb-odoo) [reply addons/mail/static/src/component_hooks/use_ref_to_model/use_ref_to_model.js:19]: Ok to rename

**Review by alexkuhn — APPROVED** [review]
@robodoo r+
